plugins/meta/firewall/README.md

```diff
@@ -133,3 +133,207 @@ of the container as shown:
+
+1. creates `originalNS` namespace
+2. adds `dummy0` interface to `originalNS` via Netlink
+3. checks that the `dummy0` interface is available in the `originalNS`
+4. creates `targetNS` namespace
+
+Upon the completion of the testing, the test harness does the following:
+
+1. closes `originalNS` namespace
+2. closes `targetNS` namespace
+
+The tests in the harness start with `It()`.
+
+Generally, a test contains a number of input arguments. In the case of
+"installs nftables rules, checks the rules exist, then cleans up on delete using v4.0.x",
+the test has the following arguments:
+
+* container id: `dummy`
+* the path to container namespace, i.e. `targetNS`
+* the name of the interface
+* the JSON payload containing a dummy request
+
+The test uses the same arguments and runs the following operations in
+`originalNS` namespace:
+
+* `cmdAdd`
+* `cmdCheck`
+* `cmdDel`
+
+The operations correspond to the following functions:
+
+| **Operation** | **Function** |
+| --- | --- |
+| `cmdAdd` | `func (nb *nftBackend) Add(conf *FirewallNetConf, result *current.Result)` |
+| `cmdCheck` | `func (nb *nftBackend) Del(conf *FirewallNetConf, result *current.Result)` |
+| `cmdDel` | `func (nb *nftBackend) Check(conf *FirewallNetConf, result *current.Result)` |
+
+The following command triggers the testing of `firewall` plugin:
+
+```bash
+sudo go test -v ./plugins/meta/firewall
+```
+
+At the outset, the test outputs dummy host and container namespaces:
+
+```
+Host Namespace: /var/run/netns/cnitest-ba94096d-68e1-90c0-5e0a-4acf3a8339cd
+Container Namespace: /var/run/netns/cnitest-762bc306-9af5-5882-af95-e011590ce8d3
+```
+
+The knowing the last part of the namespace path helps inspecting namespaces
+with `sudo ip netns exec` command. For example, the following command
+show `nftables` tables, chains, and rules.
+
+```bash
+
+$ sudo ip netns exec cnitest-ba94096d-68e1-90c0-5e0a-4acf3a8339cd nft list ruleset
+table ip filter {
+        chain FORWARD {
+                type filter hook forward priority filter; policy drop;
+                jump cnins-3-4026550857-dummy0
+        }
+
+        chain cnins-3-4026550857-dummy0 {
+                oifname "dummy0" ip daddr 192.168.100.100 ct state established,related counter packets 0 bytes 0 accept
+                iifname "dummy0" ip saddr 192.168.100.100 counter packets 0 bytes 0 accept
+                iifname "dummy0" oifname "dummy0" counter packets 0 bytes 0 accept
+        }
+}
+table ip6 filter {
+        chain FORWARD {
+                type filter hook forward priority filter; policy drop;
+                jump cnins-3-4026550857-dummy0
+        }
+
+        chain cnins-3-4026550857-dummy0 {
+                oifname "dummy0" ip6 daddr 2001:db8:100:100::1 ct state established,related counter packets 0 bytes 0 accept
+                iifname "dummy0" ip6 saddr 2001:db8:100:100::1 counter packets 0 bytes 0 accept
+                iifname "dummy0" oifname "dummy0" counter packets 0 bytes 0 accept
+        }
```

Comment:
Ideally these would be combined into one chain in a table of family `inet`. I don't think there is value in splitting them up. It can probably avoid some duplicate rules.

Reply:
> Ideally these would be combined into one chain in a table of family inet. I don't think there is value in splitting them up. It can probably avoid some duplicate rules.

@erig0, please see @mars1024  comment that he wanted a jump.

Reply:
> I don't think there is value in splitting them up. It can probably avoid some duplicate rules.

@erig0, it has a value in a sense that it is easier to delete the rule when you have a jump like this. It would be very hard to parse each rule.

Reply:
I get the jump to a dedicated chain. That 100% makes sense. I was saying I don't think you need to split the "ip" and "ip6" families. You could use "inet" which covers both IPv4 and IPv6.

Reply:
> I get the jump to a dedicated chain. That 100% makes sense. I was saying I don't think you need to split the "ip" and "ip6" families. You could use "inet" which covers both IPv4 and IPv6.

@erig0, that I could do. 👍 It makes sense because IPv4/6 are all a part of the same construct.